### PR TITLE
Fix issue 1365: support binary deserializers in CLI

### PR DESCRIPTION
**Pull request by eldanb**
### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Does your submission pass tests?

> Note: please do NOT include build files (those generate by the build-xxx commands) with your PR,

Thank you for your help in advance, much appreciated !

### CODE REVIEW

**Comment by z3dev** [comment]
@eldanb Oh! You provided the changes. Nice!

The changes are fine except for one point. Each deserializer has a specific API, and changing those now in V2 is probably not a good idea. There are other applications using those deserializers. :(

I suggest doing something like the core library does, and choose the buffer contents based on file extension.

https://github.com/jscad/OpenJSCAD.org/blob/15a50a0b349bb1db41c523e716550deb83934da3/packages/core/src/web/walkFileTree.js#L53

Also, there are some new formats coming in V3, including 3MF. So, maybe the IO library needs another list, or extend the list of 'formats'. Then there's one place that keeps track of the required 'binary' versus 'string' contents.

**Comment by eldanb** [comment]
But I didn't change the serializers' APIs. 
Their API could always receive either an array buffer or a string (that's what happens in walkFiles). Some of them would fail when they get an array buffer. That's what I fixed.

**Comment by eldanb** [comment]
Looking at binaryMimeTypes like this means that there are multiple places that encode the deserializer's expected API, which I recommend to avoid.

If you prefer we can decide that a deserializer function carries a "isBinary" property, and based on that we either load binary or text. But that too is more cumbersome, I believe, than simply declaring the deserializer's API (as it is now) as receiving "string or ArrayBuffer"

File: packages/io/amf-deserializer/src/index.js [hunk 1]
```diff
@@ -25,6 +25,7 @@ All code released under MIT license
 const version = require('../package.json').version
 const translate = require('./translate')
 const instantiate = require('./deserialize')
+const { ensureString } = require('../../io-utils')
```
Comment (z3dev) [inline packages/io/amf-deserializer/src/index.js:28]: These IO packages can be installed independently of the full JSCAD tree, so this is probably better as:

`const { ensureString } = require('@jscad/io-utils')`

That makes @jscad/io-utils a dependency for this package. A little pain.

**Review by z3dev — CHANGES_REQUESTED** [review]
@eldanb a few questions and suggestions.

File: packages/core/src/io/registerExtensions.js [hunk 2]
```diff
@@ -16,8 +16,13 @@ const registerDeserializer = (extension, fs, _require) => {
   const deserializer = deserializers[extension]
   const fileExtension = '.' + extension
   _require.extensions[fileExtension] = (module, filename) => {
-    const content = fs.readFileSync(filename, 'utf8')
-    const parsed = deserializer({ filename, output: 'geometry' }, content)
+    const fileReadResult = fs.readFileSync(filename);
+    const fileContent = fileReadResult.buffer
```
[the remark this answers is not in the document]
Comment (eldanb) [reply packages/core/src/io/registerExtensions.js:20]: There is.

As the Node API specifies (https://nodejs.org/api/buffer.html#bufbuffer), "This ArrayBuffer is not guaranteed to correspond exactly to the original Buffer. See the notes on buf.byteOffset for details."
Indeed unaware of this initially, in my tests I ran into cases where the buffer contained fragments of data from previously read files. Hence the slice. Will add a comment to that effect.

File: packages/io/amf-deserializer/src/index.js (hunk 1 above)
Comment (eldanb) [reply packages/io/amf-deserializer/src/index.js:28]: Done

File: packages/io/io-utils/ensureString.js [hunk 3]
```diff
@@ -0,0 +1,9 @@
+const ensureString = (stringOrArrayBuffer, defaultBinaryEncoding) => {
```
[the remark this answers is not in the document]
Comment (eldanb) [reply packages/io/io-utils/ensureString.js:1]: Done (Assuming your comment was to make this an explicit default parameter)

[the remark this answers is not in the document]
Comment (z3dev) [reply packages/io/io-utils/ensureString.js:1]: Correct

**Comment by eldanb** [comment]
@z3dev I've committed changes per your comments. 
Note that I've not advanced any of the package's versions, I'm assuming this is something you do as part of your release/publish routine?

**Comment by z3dev** [comment]
> @z3dev I've committed changes per your comments.
> Note that I've not advanced any of the package's versions, I'm assuming this is something you do as part of your release/publish routine?

Yeah. The version updates happen automatically via Lerna.

I'll double check.

**Comment by z3dev** [comment]
@platypii please reciew

**Review by platypii — APPROVED** [review]
The code looks fine, and I appreciate the tests.

However I have to admit I was unable to re-pro the issue? I downloaded example_project.zip from your issue #1365. But when I use the latest published cli, I don't get an error? Maybe I'm missing something?

Screenshot at 2024-11-05 11-17-26

**Comment by eldanb** [comment]
To reproduce the issue you should require() the stl. There's a sample project there.

**Comment by platypii** [comment]
Confirmed I was able to reproduce the failure on master, but success on this pr branch :+1:

**Review by z3dev — APPROVED** [review]

**Comment by z3dev** [comment]
@eldanb these changes are all set to go.

fyi, there are coding style standards. But there are probably lots of exceptions in V2.

`npm run lint`

i will roll a release this weekend.

**Pull request merged**